contracts/Strategy.sol

```diff
@@ -222,35 +261,43 @@ contract Strategy is BaseStrategy {
 
     // HELPERS //
 
-    // Manually returns lps in masterchef to the strategy. Used in emergencies.
-    function emergencyWithdrawFromMasterChef() external onlyVaultManagers {
-        _withdrawFromMasterChef(address(this));
+    function withdrawFromMasterChef(uint256 _amount, uint256 _masterChefPooId) external onlyVaultManagers {
+        _withdrawFromMasterChef(address(this), _amount, _masterChefPooId);
     }
 
     // AbandonRewards withdraws lp without rewards. Specify where to withdraw to
-    function _withdrawFromMasterChef(address _to) internal {
-        uint balanceOfBptInMasterChef = balanceOfBptInMasterChef();
-        if (balanceOfBptInMasterChef > 0) {
-            abandonRewards
-            ? masterChef.emergencyWithdraw(masterChefPoolId, address(_to))
-            : masterChef.withdrawAndHarvest(masterChefPoolId, balanceOfBptInMasterChef, address(_to));
+    function _withdrawFromMasterChef(address _to, uint256 _amount, uint256 _masterChefPoolId) internal {
+        if (_amount > 0) {
+            toggles.abandonRewards
+            ? masterChef.emergencyWithdraw(_masterChefPoolId, address(_to))
+            : masterChef.withdrawAndHarvest(_masterChefPoolId, _amount, address(_to));
         }
+    }
 
-        uint balanceOfStakeBptInMasterChef = balanceOfStakeBptInMasterChef();
-        if (balanceOfStakeBptInMasterChef > 0) {
-            abandonRewards
-            ? masterChef.emergencyWithdraw(masterChefStakePoolId, address(_to))
-            : masterChef.withdrawAndHarvest(masterChefStakePoolId, balanceOfStakeBptInMasterChef, address(_to));
-        }
+    // Withdraw the desired amount in bpt from masterchef and sell it for want
+    function _withdrawFromMasterChefAndSellBpt(uint256 amount) internal {
+        // don't try to withdraw more than what we have in masterchef
+        amount = Math.min(amount, balanceOfBptInMasterChef());
+
+        // withdraw the desired amount out of masterchef
+        _withdrawFromMasterChef(address(this), amount, masterChefPoolId);
+
+        // sell the desired amount for want
+        _sellBpt(amount, assets, tokenIndex, balancerPoolId, balanceOfBpt());
```

Comment:
This https://github.com/tonkers-kuma/strategy-ssb/pull/7/files#r801183642 could be done here instead

Reply:
It would work in almost any scenario where balanceOfBpt() == 0 before the withdrawal from masterchef. But if there is some bpt on the strategy that is not staked, those bpt will be left there instead of sold for want. That's why I didn't change it here.